File: cpp/114.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <utility>
#include <map>
#include <set>
#include <string>
#include <unordered_map>
#include <climits>
#include <numeric>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
  TreeNode *_flatten(TreeNode *root, TreeNode *end){
    if(root == nullptr)
      return end;
    root->right = _flatten(root->left, _flatten(root->right, end));
    root->left = nullptr;
    return root;
  }
  void flatten(TreeNode *root) {
    _flatten(root, nullptr);
  }
};

int main(){
  auto sol = Solution();
}

File: python/114.py
class Solution(object):
    def _flatten(self, root, end):
        if not root:
            return end
        root.right = self._flatten(root.left, self._flatten(root.right, end))
        root.left = None
        return root

    def flatten(self, root, end=None):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        self._flatten(root, None)
